### PR TITLE
[Python] Improve Concept Exercise: lists #2880

### PR DESCRIPTION
**Pull request by serenalucca**
Hi! This is my very first contribution to an open-source project, I need feedback, especially for the hints file.
Thanks in advance.

_Maintainer edit: Resolves #2880_

### CODE REVIEW

File: languages/python/exercises/concept/elyses-enchantments/.docs/hints.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+- The easiest one is the for loop, here is an example of how to use it:
+
+```python
+>>> list = [1, 2, 3, 4, 5]
+>>> for item in list:
+...     item*2
+>>>list
+[2, 4, 6, 8, 10]
+```
+
```
Comment (BethanyG) [inline languages/python/exercises/concept/elyses-enchantments/.docs/hints.md:54]: I really like this example, and the `hints.md` file overall.  But I am a bit "on the fence".  These are _hints_, and so I don't know if we want specific code examples.  @valentin-p @cmccandless  your thoughts here & for the rest of the file?

File: languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md [hunk 2]
```diff
@@ -0,0 +1,242 @@
+In Python, a [list][list] is a mutable collection of items in _sequence_. Like most collections (_see the built-ins [`tuple`][tuple], [`dict`][dict] and [`set`][set]_), lists can hold reference to any (or multiple) data type(s) - including other lists. Like any [sequence][sequence type], items are referenced by 0-based index number, and can be copied in whole or in part via _slice notation_. Lists support all [common sequence operations][common sequence operations], as well as [mutable sequence operations][mutable sequence operations] like `.append()` and `.reverse()`. They can be iterated over in a loop by using the `for item in` construct.
```
Comment (BethanyG) [inline languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md:1]:
```suggestion
In Python, a [list][list] is a mutable collection of items in _sequence_. Like most collections (_see the built-ins [`tuple`][tuple], [`dict`][dict] and [`set`][set]_), lists can hold reference to any (or multiple) data type(s) - including other lists. Like any [sequence][sequence type], items are referenced by 0-based index number, and can be copied in whole or in part via _slice notation_. Lists support all [common sequence operations][common sequence operations], as well as [mutable sequence operations][mutable sequence operations] like `.append()` and `.reverse()`. They can be iterated over in a loop by using the `for item in <list>` construct.
```

File: languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+Under the hood, lists are implemented as [dynamic arrays][dynamic array] -- similar to Java's [`Arraylist`][arraylist] type. Lists are most often used to store groups of similar data (_strings, numbers, sets etc._) of unknown length (_the number of entries may arbitrarily expand or shrink_). Accessing items in a list, checking for membership via `in`, or appending items to the "right-hand" side of a list are all very efficient. Appending to the "left-hand" side or inserting into the middle of a list is much _less_ efficient because it requires shifting items to keep them in sequence.
```
Comment (BethanyG) [inline languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md:3]:
```suggestion
Under the hood, `lists` are implemented as [dynamic arrays][dynamic array] -- similar to Java's [`Arraylist`][arraylist] type. Lists are most often used to store groups of similar data (_strings, numbers, sets etc._) of unknown length. Accessing items, checking for membership via `in`, or appending items to the "right-hand" side of a list are all very efficient. Appending to the "left-hand" side or inserting into the middle of a list is much _less_ efficient because it requires shifting items to keep them in sequence.
```

File: languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+>>> multiple_code_points_string = list('अभ्यास')
+['अ', 'भ', '्', 'य', 'ा', 'स']
+
+"""
+The iteration default for dictionaries is over the keys.
+"""
+source_data = {"fish": "gold", "monkey": "brown"}
+>>> multiple_elements_dict_1 = list(source_data)
+['fish', 'monkey']
+
```
Comment (BethanyG) [inline languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md:87]:
```suggestion
```

File: languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+Because the constructor will only take _iterables_ (or nothing) as arguments, objects that are _not_ iterable will throw a type error. Consequently, it is much easier to create a one-item list via the literal method.
```
Comment (BethanyG) [inline languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md:88]:
```suggestion
Because the `list` constructor will only take _iterables_ (or nothing) as arguments, objects that are _not_ iterable will throw a type error. Consequently, it is much easier to create a one-item list via the literal method.
```

File: languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+
+>>> cubes_to_1000 = []
+
+>>> for number in range(11):
+...    cubes_to_1000.append(number**3)
+
+>>> print(cubles_to_1000)
+[0, 1, 8, 27, 64, 125, 216, 343, 512, 729, 1000]
+```
+
```
Comment (BethanyG) [inline languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md:195]:
```suggestion
```

File: languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+>>> first_one
+["George", 5, "cat", "Tabby"]
+
+# this loops through the first list and appends it's items to the end of the second list
+>>> for item in first_one:
+>>>      second_one.append(item)
+...
+>>> print(second_one)
+["George", 5, "cat", "Tabby"]
+```
```
Comment (BethanyG) [inline languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md:227]:
```suggestion
```

File: languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+
+## Related data types
+
+Lists are often used as _stacks_ and _queues_ -- although their underlying implementation makes prepending and inserting slow. The [collections][collections] module offers a [deque][deque] variant optimized for fast appends and pops from either end that is implemented as a [doubly linked list][doubly linked list]. Nested lists are also used to model small _matrices_ -- although the [Numpy][numpy] and [Pandas][pandas] libraries are much more robust for efficient matrix and tabular data manipulation. The collections module also provides a `UserList` type that can be customized to fit specialized list needs.
```
Comment (BethanyG) [inline languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md:231]:
```suggestion
```

File: languages/python/exercises/concept/elyses-enchantments/.meta/design.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
+## Goal
+
+The goal of this exercise is to teach the basics of the `List` (sequence type) data type in Python. The exercise will walk the student through how to create lists via various methods, iterate through lists via looping, and access items in various lists through indexing and forming/accessing list slices.
```
Comment (BethanyG) [inline languages/python/exercises/concept/elyses-enchantments/.meta/design.md:3]:
```suggestion
The goal of this exercise is to teach the basics of the `list` data type in Python. The exercise will walk the student through how to create lists via various methods, iterate through lists via looping, and access items in various lists through indexing and forming/accessing list slices.
```

File: languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md (hunk 6 above)
Comment (BethanyG) [inline languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md:195]:
```suggestion
```

File: languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+>>> breakfast_foods.insert(3,"Bacon")
+>>> print(breakfast_foods)
+['Oatmeal', 'Fruit Salad', 'Eggs', 'Bacon', 'Toast']
+
+
+#adding coffee in the first position
+>>> breakfast_foods.insert(0, "Coffee")
+>>> print(breakfast_foods)
+['Coffee', 'Oatmeal', 'Fruit Salad', 'Eggs', 'Bacon', 'Toast']
+```
```
Comment (BethanyG) [inline languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md:155]:
```suggestion
```

File: languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+>>> one_element = list(16)
+Traceback (most recent call last):
+   File "<stdin>", line 1, in <module>
+   TypeError: 'int' object is not iterable
+
+>>> one_element_from_iterable = list((16,))
+[16]
+````
+
+## Accessing elements
```
Comment (BethanyG) [inline languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md:101]: I didn't realize that this exercise used slicing in its solution when I wrote this intro (apologies!), so I think that we'll need to add some explanation for _slice notation_ (`[start:stop:step]`) , as well as some explanation of what a slice is (_a copy of a portion of the original list_).

**Comment by BethanyG** [comment]
@serenalucca - **Good work on this!**  Thank you so much for submitting it.  🎉   I've made a few suggestions above - most of them are around reducing/shortening the content in `introduction.md`.  

The `introduction.md` file is intended to only introduce concepts needed for a student to complete the exercise -- so the extra list methods and tips on slice assignment aren't needed in this context.  Conversely, we'll need to add in some tips on slicing since this exercise does use list slices.

File: languages/python/exercises/concept/elyses-enchantments/.docs/hints.md (hunk 1 above)
Comment (serenalucca) [reply languages/python/exercises/concept/elyses-enchantments/.docs/hints.md:54]: Thank you, I tried my best but I wasn't sure how precise I should be because I know that they would only use the hint in last resort but at the same time this subject is not too difficult and it is well explained in the python docs.

File: languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md (hunk 11 above)
Comment (serenalucca) [reply languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md:95]: I added the explanation of the slice notation in the  intro and about file, I tried my best to make it clear but I don't know if my writing style is adapted to the style of the file. Can you please check that?
Thank you very very much for all the feedback!

File: languages/python/exercises/concept/elyses-enchantments/.docs/hints.md (hunk 1 above)
Comment (valentin-p) [reply languages/python/exercises/concept/elyses-enchantments/.docs/hints.md:54]: I agree that explaining in plain English would be more suitable.

See for example: [dicts hints.md](https://github.com/exercism/v3/blob/master/languages/python/exercises/concept/inventory-management/.docs/hints.md)

**Review by valentin-p — COMMENTED** [review]
It's a really good pull request here and great job
 Thanks for that!

You may need to run prettier on each .MD file that you created/updated manually for the checks to pass.

File: languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+>>> one_element = ["Guava"]
+["Guava"]
+
+>>> elements_separated_with_commas = ["Parrot", "Bird", 334782]
+["Parrot", "Bird", 334782]
+```
+
+For readability, line breaks can be used when there are many elements or nested data structures within a list:
+
+```python
```
[the remark this answers is not in the document]
Comment (serenalucca) [reply languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md:24]: I shortened the list in that file but I left it whole in the about doc, is it alright?

File: languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+#the set is unpacked and each element is added
+>>> multiple_elements_set = list({2, 3, 5, 7, 11})
+[2,3,5,7,11]
+```
+
+Results when using a list constructor with a string or a dict may be surprising:
+
+````python
+
+#string elements (Unicode code points) are iterated through and added *individually*
```
[the remark this answers is not in the document]
Comment (serenalucca) [reply languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md]: Done! I hope that I didn't forget any and did it well enough.

File: languages/python/exercises/concept/elyses-enchantments/.docs/hints.md (hunk 1 above)
Comment (serenalucca) [reply languages/python/exercises/concept/elyses-enchantments/.docs/hints.md:42]: I updated the file, could you check if it is better now? Thanks

Comment (valentin-p) [reply languages/python/exercises/concept/elyses-enchantments/.docs/hints.md:42]: Thank you for all the work, including fixing little things that you did not write yourself. That's perfect.

**Review by valentin-p — APPROVED** [review]
LGTM. Thanks for all the fixes and the added text. That helps a lot. We are closer to have a great collection of concept exercises, ready to be solved.

File: languages/python/concepts/lists/about.md [hunk 14]
```diff
@@ -112,21 +112,36 @@ Items inside lists (_like the sequence types `string` and `tuple`_), can be acce
...
-#Oatmeal is at index 0 or index -4
+# Oatmeal is at index 0 or index -4.
 >>> first_breakfast_food = breakfast_foods[0]
 'Oatmeal'
 
 >>> first_breakfast_food = breakfast_foods[-4]
 'Oatmeal'
 ```
 
+You can access a portion of the list with the slice notation `[start:stop]`. The slice of a list from start to stop is defined as the sequence of items with index `i` such that `start <= i < stop`.
```
Comment (BethanyG) [inline languages/python/concepts/lists/about.md:123]:
```suggestion
You can access a portion of a list with _slice notation_ (`[start:stop]`). A _slice_  is defined as the sequence of items in a list at position `index` such that `start <=  index < stop`.  Slicing a `list` returns a copy of the data.
```

File: languages/python/concepts/lists/about.md [hunk 15]
```diff
@@ -112,21 +112,36 @@ Items inside lists (_like the sequence types `string` and `tuple`_), can be acce
...
+You can also get a slice of a list from `start` to `stop` with step `step` with the notation `[start:stop:step]`.
+A slice of a list is a copy of a portion of the original list so you do not modify the original but you get a new list with the elements you asked for.
```
Comment (BethanyG) [inline languages/python/concepts/lists/about.md:125]:
```suggestion
You can also slice a list using a `step` parameter with the notation `[start:stop:step]`.  Using a `step` will "skip over" or filter the list elements (_for example, a `step` of 2 will be every other element in the range_).
```

File: languages/python/concepts/lists/about.md [hunk 16]
```diff
@@ -112,21 +112,36 @@ Items inside lists (_like the sequence types `string` and `tuple`_), can be acce
...
+
+```python
+
+>>> colors = ["Red", "Purple", "Green", "Yellow", "Orange", "Pink", "Blue", "Grey"]
+
+>>> middle_colors = colors[2:6]
+["Green", "Yellow", "Orange", "Pink"]
+
+>>> primary_colors = colors[0:len(colors):3]
```
Comment (BethanyG) [inline languages/python/concepts/lists/about.md:134]:
```suggestion
>>> primary_colors = colors[0::3]
```

File: languages/python/concepts/lists/about.md [hunk 17]
```diff
@@ -112,21 +112,36 @@ Items inside lists (_like the sequence types `string` and `tuple`_), can be acce
...
 
+You can access a portion of the list with the slice notation `[start:stop]`. The slice of a list from start to stop is defined as the sequence of items with index `i` such that `start <= i < stop`.
+You can also get a slice of a list from `start` to `stop` with step `step` with the notation `[start:stop:step]`.
+A slice of a list is a copy of a portion of the original list so you do not modify the original but you get a new list with the elements you asked for.
+
+```python
+
+>>> colors = ["Red", "Purple", "Green", "Yellow", "Orange", "Pink", "Blue", "Grey"]
+
+>>> middle_colors = colors[2:6]
```
Comment (BethanyG) [inline languages/python/concepts/lists/about.md:132]:
```suggestion
# If there is no step parameter, the step is assumed to be 1
>>> middle_colors = colors[2:6]
```

File: languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md (hunk 11 above)
Comment (BethanyG) [reply languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md:85]: I added a few small things.  Mostly around trying to clarify slicing - which I think is hard for people to understand on first pass.  I am not so sure I "improved" things, so feel free to reject my suggestions!

File: languages/python/concepts/lists/about.md (hunk 17 above)
Comment (serenalucca) [reply languages/python/concepts/lists/about.md:132]: Thanks you for thé suggestions ! Should I also apply them to the introduction file?

File: languages/python/concepts/lists/about.md (hunk 16 above)
Comment (BethanyG) [reply languages/python/concepts/lists/about.md]: Just a note -- all three variants work -- (`colors[0:len(colors):3]`, `colors[0::3]`, and `colors[0::3]`) but since the runtime is keeping track of the start/stop of the list internally, calling `len(colors)` just adds an extra method call, and adding the 0 is just extra typing.  😉

File: languages/python/concepts/lists/about.md (hunk 17 above)
Comment (BethanyG) [reply languages/python/concepts/lists/about.md:132]: ooh!  Yes, please.  I haven't (yet) read that far!  😄

**Comment by BethanyG** [comment]
@serenalucca  - **Awesome work** on this!  ⭐ .  I can go ahead and run `format` if you are done editing.  Then we can merge!  YAY.  Thank you for all of your hard work on this, it is much appriciated.

**Review by BethanyG — APPROVED** [review]

File: languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md (hunk 11 above)
Comment (serenalucca) [reply languages/python/exercises/concept/elyses-enchantments/.docs/introduction.md:85]: I find your comments on the code very helpful added those from about to the into doc

**Comment by serenalucca** [comment]
I did a last commit to add the comments to the introduction file. Do I have to run prettier before you run format or does the format command applies prettier on the files?
If it is alright for you, you can merge ! Thank you very very much for all the help on this PR!

**Comment by valentin-p** [comment]
If you can run prettier manually than would be awesome, because it seems like format isn't helping much, we will need to check that. We can merge only if the "CI/Verify formatting" is green.

**Comment by BethanyG** [comment]
@serenalucca  - I think we are good! 🌈  🌟 

But I **just** noticed that you made your changes from `master` on your fork.  It's fine for this PR, and I don't think there are any issues merging...but we think that's why `/format` refused to run. 😄 

So in the future (_and we'd love if you want to do more PRs for us!_) we recommend doing any changes on a branch in your fork.  That will separate the changes better, and will avoid a naming confusion.

**Comment by serenalucca** [comment]
Thank you very very much for accepting the PR! 🎉 I hope that I will have the chance to do this again someday. Thank you for the tip, I got a little confused as it was my first PR. Thank you for being all the reviews @BethanyG and @valentin-p, I had a great time working  on this! 😄
